libcontainer/standard_init_linux.go

```diff
@@ -233,13 +234,26 @@ func (l *linuxStandardInit) Init() error {
 	// user process. We open it through /proc/self/fd/$fd, because the fd that
 	// was given to us was an O_PATH fd to the fifo itself. Linux allows us to
 	// re-open an O_PATH fd through /proc.
-	fifoPath := "/proc/self/fd/" + strconv.Itoa(l.fifoFd)
-	fd, err := unix.Open(fifoPath, unix.O_WRONLY|unix.O_CLOEXEC, 0)
-	if err != nil {
-		return &os.PathError{Op: "open exec fifo", Path: fifoPath, Err: err}
-	}
-	if _, err := unix.Write(fd, []byte("0")); err != nil {
-		return &os.PathError{Op: "write exec fifo", Path: fifoPath, Err: err}
+	ch := make(chan error, 1)
+	go func() {
+		fifoPath := "/proc/self/fd/" + strconv.Itoa(l.fifoFd)
+		fd, err := unix.Open(fifoPath, unix.O_WRONLY|unix.O_CLOEXEC, 0)
+		if err != nil {
+			ch <- &os.PathError{Op: "open exec fifo", Path: fifoPath, Err: err}
+		}
+		if _, err := unix.Write(fd, []byte("0")); err != nil {
+			ch <- &os.PathError{Op: "write exec fifo", Path: fifoPath, Err: err}
+		}
+		ch <- nil
+	}()
+
+	select {
+	case chErr := <-ch:
+			if chErr != nil {
+					return chErr
+			}
+	case <-time.After(120 * time.Second):
```

Comment:
To be honest, it is possible to call runC start after 2mins. And I don't think the runC-init process is leaky because contained got killed. 

It is possible if the user uses containerd-shim-v1, because the resource cleanup requires shim-v1 server is alive. But the community deprecates containerd-shim-v1 and suggest that user uses v2. In v2 design, the containerd can make sure the resource created by shim will be cleanup, including runC-init process.

IMO, the runC-init process should be cleanup by caller, instead of subside.

Reply:
I also think this is trying to fix the problem on the wrong side. By design, runc waits indefinitely, and if I remember correctly the way to clean it up is to call runc kill.

Reply:
shmi-v2 also doesn't fix the init leak if the containerd exits unexpectedly like a segmentfault

Reply:
@kamizjw could you mind to share how to reproduce this? If it is bug in containerd, I would like to fix it in containerd site.